docs(sca): documentation related to the new SCA mode (#2455)
bonitasoft/bonita-doc: modules/build-run/pages/sca-user-flows.adoc

= Build and run an Application
:experimental: // activate the 'menu' macro

> :description: User flows possible in Application mode

Author: Application mode should not be used

== User flows in Application mode

=== Creating a new project in Studio using Application mode

[WARNING]
====
In order to generate the docker image, it's mandatory to have a maven repository working to get all the dependencies, including the latest Application runtime version.
====

. Run Bonita Studio {bonitaVersion} and create a new project.

. Configure the project environment

. Validate the project
.. To work with Application mode, you need to explicitly mapping a form for each interaction form (instantiation and manual task)

> . Admin and user app is now available in extension, if you want to deploy it on the final bundle with Application mode, you need to add it manually

Author: ```suggestion
. Admin and user app are now available as extensions, the Bonita Admin App is added by default on new projects, if you want to include the Bonita User App in your Application, you need to add it manually from the Marketplace.
```

Reviewer: Bonita Admin App is added only in Enterprise edition I guess

.. Go to the Extension view

> .. Open Marketplace and filter on “Application type”

Author: ```suggestion
.. Open Marketplace and filter on “Application” type
```

> .. You can now add the default application as you wish on the final bundle

Author: ```suggestion
.. You can now add the Bonita provided applications to include in your application.
```

> . At the root of the project folder, run `mvn clean package -Pbundle,docker "-Dbonita.environment=production"` 

Author: Link to the build-application article instead. Choose a single packaging, either Docker image or Tomcat bundle. Not both.

> . Find the docker image name with `docker images` to be sure it's being created.

Author: Overkill, maybe precise that Docker must be installed as prerequisite before building with the -Pdocker argument.

> . Run the docker image with command (adapting licenses folder if needed) `docker run --rm --name MyProjectName -h MyProjectName -e -v ~/lic/:/opt/bonita_lic/ -d -p 9000:8080 MyProjectNameImage-production:0.0.1`

Author: Link to the run-application articale instead.
The license folder mapping and hostname declaration is specific to sp edition.

Reviewer: In case you mention a command like this, please use a 'source' section such as:

//[source,shell]
//----
//docker run ...
//----

This is more readable and easier to copy the command with a single click

> . Inspect logs with the command `docker logs` 

Author: Just removes the `-d` option above to simplify the use case. 
Detailed explanation on how to access container logs in Docker can be described in the run-application article.

> . Go to http://localhost:9000/bonita[http://localhost:9000/bonita], launch a process

Author: What process and how ? First page will be the login page...

> . Your application is working! 

Author: Precise that it is running with the h2 embedded database in this case.

[NOTE]
====

> In case you need to clean your docker: Stop and delete the docker container with docker rm -f MyProjectName and remove the docker image with docker rmi -f MyProjectNameImage-production:0.0.1

Author: Not sure if this is relevanr here

====

> === Migrate an existing project in Studio to Application mode

Author: ```suggestion
=== Update an existing project using Bonita Studio
```

> . Run Bonita Studio {bonitaVersion} and import from .bos or Git your project to migrate

Author: ```suggestion
. Run Bonita Studio {bonitaVersion} and import from .bos or clone a Git project to update
```

> . Migrate your project

Author: ```suggestion
. Update your project
```

. Validate your project

> .. To work with Application mode, you need to explicitly mapping a form for each interaction form (instantiation and manual task)

Author: ```suggestion
.. To have a working Application, you must explicitly map all forms (Process instantiation and Task forms).
```

> . Admin and user app is now available in extension, if you want to deploy it on the final bundle with Application mode, you need to add it manually

Author: ```suggestion
. Admin and user app are now available as extensions, the Bonita Admin App is added by default on updated projects, if you want to include the Bonita User App in your Application, you need to add it manually from the Marketplace.
```

.. Go to the Extension view

> .. Open Marketplace and filter on “Application type”

Author: ```suggestion
.. Open Marketplace and filter on “Application” type
```

> .. You can now add the default application as you wish on the final bundle

Author: ```suggestion
.. You can now add the Bonita provided applications to include in your application.
```

> . At the root of the project folder, run `mvn clean package -Pbundle,docker "-Dbonita.environment=qualification"`

Author: Same as above

> . To run the docker images build in previous step, you need to run: `docker run --rm --name MyProjectName -h MyProjectName -v ~/lic/:/opt/bonita_lic/ -d -p 9000:8080 MyProjectNameImage-qualification:0.0.1`

Author: Same as above

>  .. If you need custom-permissions (in restApiExtensions as example), you can create another volume and map this volume with a local folder (with the right custom-mapping-permission file) with `-v ~/bonita-tenant-template-portal/:/opt/bonita/setup/platform_conf/initial/tenant_template_portal/`

Author: Too specific ? To be documented in the Run application article ?

[NOTE]
====
More information about bonita docker configuration can be found xref:runtime:bonita-docker-installation.adoc[here]
====

> . Go to http://localhost:9000/bonita[http://localhost:9000/bonita], launch a process

Author: First page will be the login page...
